Tests/designspaceLib/designspace_test.py

```diff
@@ -0,0 +1,870 @@
+    r2.name = "named.rule.2"
+    r2.conditions.append(dict(name='aaaa', maximum=500))
+    r2.subs.append(("b", "b.alt"))
+
+    assert evaluateRule(r2, dict(aaaa = 0)) == True
+    assert evaluateRule(r2, dict(aaaa = -500)) == True
+    assert evaluateRule(r2, dict(aaaa = 1000)) == False
+
+    # rule with only a minimum
+    r3 = RuleDescriptor()
+    r3.name = "named.rule.3"
+    r3.conditions.append(dict(name='aaaa', minimum=500))
+    r3.subs.append(("c", "c.alt"))
+
+    assert evaluateRule(r3, dict(aaaa = 0)) == False
+    assert evaluateRule(r3, dict(aaaa = 1000)) == True
+    assert evaluateRule(r3, dict(bbbb = 1000)) == True
+
+    # rule with only a minimum, maximum in separate conditions
+    r4 = RuleDescriptor()
+    r4.name = "named.rule.4"
+    r4.conditions.append(dict(name='aaaa', minimum=500))
+    r4.conditions.append(dict(name='bbbb', maximum=500))
+    r4.subs.append(("c", "c.alt"))
+
+    assert evaluateRule(r4, dict()) == True  # is this what we expect though?
+    assert evaluateRule(r4, dict(aaaa = 1000, bbbb = 0)) == True
+    assert evaluateRule(r4, dict(aaaa = 0, bbbb = 0)) == False
+    assert evaluateRule(r4, dict(aaaa = 1000, bbbb = 1000)) == False
+
+    a1 = AxisDescriptor()
+    a1.minimum = 0
+    a1.maximum = 1000
+    a1.default = 0
+    a1.name = "aaaa"
+    a1.tag = "aaaa"
+    b1 = AxisDescriptor()
+    b1.minimum = 2000
+    b1.maximum = 3000
+    b1.default = 2000
+    b1.name = "bbbb"
+    b1.tag = "bbbb"
+    doc.addAxis(a1)
+    doc.addAxis(b1)
+    assert doc._prepAxesForBender() == {'aaaa': {'map': [], 'name': 'aaaa', 'default': 0, 'minimum': 0, 'maximum': 1000, 'tag': 'aaaa'}, 'bbbb': {'map': [], 'name': 'bbbb', 'default': 2000, 'minimum': 2000, 'maximum': 3000, 'tag': 'bbbb'}}
+
+    assert doc.rules[0].conditions == [{'minimum': 0, 'maximum': 1000, 'name': 'aaaa'}, {'minimum': 0, 'maximum': 3000, 'name': 'bbbb'}]
+
+    assert doc.rules[0].subs == [('a', 'a.alt')]
+
+    doc.normalize()
+    assert doc.rules[0].name == 'named.rule.1'
+    assert doc.rules[0].conditions == [{'minimum': 0.0, 'maximum': 1.0, 'name': 'aaaa'}, {'minimum': 0.0, 'maximum': 1.0, 'name': 'bbbb'}]
+
+    doc.write(testDocPath)
+    new = DesignSpaceDocument()
+
+    new.read(testDocPath)
+    assert len(new.axes) == 4
+    assert len(new.rules) == 1
+    new.write(testDocPath2)
+
+
+def __removeAxesFromDesignSpace(path):
+    # only for testing, so we can make an invalid designspace file
+    # without making the designSpaceDocument also support it.
+    f = open(path, 'r', encoding='utf-8')
+    d = f.read()
+    f.close()
+    start = d.find("<axes>")
+    end = d.find("</axes>")+len("</axes>")
+    n = d[0:start] + d[end:]
+    f = open(path, 'w', encoding='utf-8')
+    f.write(n)
+    f.close()
+
+
+@pytest.fixture
+def invalid_designspace():
+    p = "testCheck.designspace"
+    __removeAxesFromDesignSpace(p)
+    yield p
+
+
+@pytest.mark.xfail(reason="The check method requires MutatorMath")
```

Comment:
I disabled this test since it requires the mutatorMath functionality. Should I leave it or remove altogether?

Reply:
leave it as xfail until we reimplment that piece of code from mutatormath in here